microcosm_pubsub/chain/chain.py

```diff
@@ -0,0 +1,65 @@
+from marshmallow import ValidationError
+
+from microcosm_pubsub.chain.context_decorators import (
+    get_from_context,
+    save_to_context,
+    save_to_context_by_func_name,
+    temporary_replace_context_keys,
+)
+
+
+class Chain:
+    """
+    Structure that contains callable functions that shares the same context.
+
+    """
+    def __init__(self, *args):
+        """
+        Build a chain that can be resolved multiple times
+        :param *args: callable functions, Chain statemnts or other Chains
+
+        """
+        self.pieces = args
+
+        self.context_decorators = [
+            # Order matter - get_from_context should be first
+            get_from_context,
+            temporary_replace_context_keys,
+            save_to_context,
+            save_to_context_by_func_name,
+        ]
+
+    def __call__(self, context=None):
+        """
+        Resolve the chain and return the last chain function result
+        :param context: optional argument - dictionary to share the context
+
+        """
+        return self.resolve(context=context)
+
+    def resolve(self, context=None, **kwargs):
```

Comment:
Notice - the chain obj knows nothing about the context - it's a separate object.